File: Lab05Cpp/Sensor.cpp
#include "Sensor.hpp"
#include <iostream>

Sensor::Sensor(const std::string& id,
               const std::string& name,
               const std::string& unit)
    : m_id(id)
    , m_name(name)
    , m_unit(unit)
{
}

Sensor::Sensor(const Sensor& other)
    : m_id(other.m_id)
    , m_name(other.m_name)
    , m_unit(other.m_unit)
    , m_lastValue(other.m_lastValue)
    , m_status(other.m_status)
{
}

bool Sensor::CheckStatus() const
{
    return m_status == "OK";
}

void Sensor::Calibrate()
{
    std::cout << "Calibrating sensor " << m_id << " (" << m_name << ")\n";
}

Sensor& Sensor::SetStatus(const std::string& status)
{
    this->m_status = status; // Явное использование this.
    return *this;            // Возвращаем *this для цепочек вызовов.
}

File: Lab04/Lab04/Lab04Monitoring/Sensor.cs
// Абстрактный базовый класс для любых датчиков.
// Здесь:
//   * используем свойства вместо публичных полей;
//   * демонстрируем разумное использование this в методе SetStatus.
using System;

namespace Lab04Monitoring
{
    public abstract class Sensor
    {
        public string Id { get; }
        public string Name { get; }
        public string Unit { get; }

        // Последнее значение и статус датчика храним в свойствах.
        public float LastValue { get; protected set; }
        public string Status { get; private set; } = "OK";

        protected Sensor(string id, string name, string unit)
        {
            Id = id;
            Name = name;
            Unit = unit;
        }

        // Абстрактное чтение значения – каждый потомок реализует по‑своему.
        public abstract float ReadValue();

        public virtual bool CheckStatus() => Status == "OK";

        public virtual void Calibrate()
        {
            Logger.Instance.Log($"Sensor {Id} calibrated.");
        }

        // Метод использует this, возвращает сам объект.
        // Это позволяет писать цепочки вида: sensor.SetStatus("ERROR").Calibrate();
        public Sensor SetStatus(string status)
        {
            this.Status = status; // Явное использование this для наглядности.
            return this;
        }
    }
}
